### ISSUE TITLE
a11y issues/improvements

### ISSUE THREAD

**Issue by mayank99** [withastro/docs#620]
List of accessibility issues I ran into when using the docs site:
- The table of contents ("on this page") section is at the very end of the document. This also includes the theme switcher.
  - I'm guessing this was done to avoid having to tab through the whole list before getting to the main content, but it arguably serves no purpose anymore. One way to fix this would be add another skip link dedicated for this table. 
- "Astro" and "Docs" links at the top left lack visible focus indicator.
- "Aside" components (e.g. alert or informational) use the `<blockquote>` element which is semantically incorrect.
  - Some of them are also italicized not just stylistically but also semantically using the `<em>` element, which seems wrong.
- Code blocks are focusable, even though there is no interaction within them.

Also, some of the borders/underlines get messed up when using Windows forced high contrast mode. e.g. you can't tell which tab is active.
![](https://user-images.githubusercontent.com/9084735/169622849-7de3885f-8679-4b7e-b41b-8e057ca869d9.png)
![](https://user-images.githubusercontent.com/9084735/169622824-bf8132fb-7295-4514-8ac2-7b0c5ab87cd1.png)

There are also some forced contrast issues in the search popup but I'm guessing the UI is not customizable so I'd have to file that under the algolia repo.

---

P.S. I would love to create PRs to fix some of these things myself.

**Comment by delucis** [comment]
Awesome! Thanks so much or putting together this list 💜

- I suspect the table of contents DOM position was just to have the three columns in source order as we switch it out to before the main article content on mobile. So tab order actually varies depending on viewport width.

- The code blocks are made focusable here: https://github.com/withastro/docs/blob/3ffa4f660f7b24245bfc39feeab279acf5eba84f/public/make-scrollable-code-focusable.js
   I think the idea is that scrollable elements should be able to receive focus and some code blocks overflow so need this?

- #401 also suggests replacing `<blockquote>` usage.

Happy to look at PRs and thanks again for the attention to detail!

**Comment by mayank99** [comment]
@delucis I've created #692 to fix some of the issues. It's not much but it's a start. I might tackle forced colors mode in a separate PR.

> I suspect the table of contents DOM position was just to have the three columns in source order as we switch it out to before the main article content on mobile. So tab order actually varies depending on viewport width.

I've fixed the order on desktop but left it as-is on mobile, so now it comes before the main content on both. We might be possibly able to use the same DOM element but I left the duplication for now.

>   I think the idea is that scrollable elements should be able to receive focus and some code blocks overflow so need this?

Oh I hadn't considered that scenario. I suppose we could use some JS to only make it focusable when actually needed, but the current solution is simpler without too much of a downside 👍

> #401 also suggests replacing `<blockquote>` usage.

I like your suggestion! I will cross it off my list then.

**Comment by sarah11918** [comment]
@mayank99 We did recently merge our new Aside component, so if you haven't looked at the docs recently, please do and let us know if you have any further comments on that!

Otherwise, are there still outstanding issues here you're hoping to PR? Or, is can this issue be closed? (Of course, you're welcome to open up new ones!)

**Comment by delucis** [comment]
Thanks for checking in on that @mayank99! The yellow is always an issue — thought I’d checked all those ratios but evidently missed that one!

I _think_ the `aria-hidden` is used correctly in this case as we use an `aria-label` with the same contents as that text node on the `<aside>` instead, but happy to revisit that if you see any screenreader issues there (our macOS VoiceOver and NVDA tests seemed to work as expected).

**Comment by delucis** [comment]
So I’m trying to reproduce those contrast numbers @mayank99 and getting this in Firefox:

<img width="174" alt="image" src="https://user-images.githubusercontent.com/357379/174486661-59114c55-b1f2-452a-86d4-8002e9daf7c0.png">

What tool are you using to check contrast?

**Comment by delucis** [comment]
Ah — I think I see the issue: the colours reported in your tooltip screenshot are not actually the colours on screen. I pasted the hex values over here — https://colourcontrast.cc/ffbd2e/ffe7b3 — and it reports a similar ratio of 1.38, but the background is semi-transparent and so much darker than the raw hex value suggests. Looks like we’re OK on that front although I’m seeing some contrast failures for bullet points and `<code>` snippets I’ll try and fix.

**Comment by mayank99** [comment]
Oh wow, yeah 1.38 doesn't even look right at first glance. I recently started usinG [Visbug](https://visbug.web.app/). I might open an issue there.

I double checked using an eyedropper to get the actual visible color and it passes with 4.95 👍

As for the `aria-label`, is there any reason to not just rely on the contents of the `<p>` instead? I think the `aria-hidden` should only be applied on the icon.

**Comment by delucis** [comment]
Oh wow — cool tool! Transparency seems to trip up a lot of these tools. Even Axe just fails when it finds transparent elements and asks you to check manually. I’ve only found the built-in browser tools to work reasonably well. (In our case it’s also gradient + semi-transparent background, so it’s a lot for the tool to figure out.

Here’s the PR review thread where I switched from `aria-labelledby` to `aria-label` + `aria-hidden`: https://github.com/withastro/docs/pull/761#discussion_r895005784

TLDR:

`<aside>` being a sectioning element should have a title. Options for that are:

- adding a heading element (`<h1>`?)
- using `aria-labelledby` to point to a text node as the section label
- using `aria-label` to specify it explicitly

I wasn’t sure we could use a heading because I wouldn’t easily know which level it should be, so used `aria-labelledby` initially. To do that, I generated a random ID for the `<p>` to use, but there was push back against that so switched to `aria-label` and hid the actual text to avoid duplicate content.

**Comment by mayank99** [comment]
> so used aria-labelledby initially. To do that, I generated a random ID for the <p> to use, but there was push back against that so switched to aria-label

Ah ok, thanks for clarifying, because I was going to suggest that exact approach (which I think is completely valid as long as the ids are unique).

But just to double confirm, I tested it in NVDA just now and the current `aria-label` seems to be announced fine, so we should be good!

Great work on the `Aside` component overall 👍

**Comment by delucis** [comment]
Cool — thanks for testing again for us, really appreciate it 💜 

I think the opposition to random IDs was only that it means build output is no longer deterministic, so can make testing harder (not that we’re actually doing that testing, but I was happy to leave the door open to it if this change helps and doesn’t hurt a11y).

I think that just leaves the high contrast mode right? Look forward to getting that addressed too.
